storage/include/rocksdb/native_client.h

```diff
@@ -0,0 +1,542 @@
+template <typename KeySpan, typename ValueSpan>
+void NativeClient::put(const std::string &cFamily, const KeySpan &key, const ValueSpan &value) {
+  auto s =
+      client_->dbInstance_->Put(::rocksdb::WriteOptions{}, columnFamilyHandle(cFamily), toSlice(key), toSlice(value));
+  throwOnError("put() failed"sv, std::move(s));
+  client_->storage_metrics_.tryToUpdateMetrics();
+}
+
+template <typename KeySpan>
+std::optional<std::string> NativeClient::get(const std::string &cFamily, const KeySpan &key) const {
+  auto value = std::string{};
+  auto s = client_->dbInstance_->Get(::rocksdb::ReadOptions{}, columnFamilyHandle(cFamily), toSlice(key), &value);
+  if (s.IsNotFound()) {
+    return std::nullopt;
+  }
+  throwOnError("get() failed"sv, std::move(s));
+  client_->storage_metrics_.tryToUpdateMetrics();
+  return value;
+}
+
+template <typename KeySpan>
+void NativeClient::del(const std::string &cFamily, const KeySpan &key) {
+  auto s = client_->dbInstance_->Delete(::rocksdb::WriteOptions{}, columnFamilyHandle(cFamily), toSlice(key));
+  throwOnError("del() failed"sv, std::move(s));
+  client_->storage_metrics_.tryToUpdateMetrics();
+}
+
+template <typename KeySpan, typename ValueSpan>
+void NativeClient::put(const KeySpan &key, const ValueSpan &value) {
+  return put(defaultColumnFamily(), key, value);
+}
+
+template <typename KeySpan>
+std::optional<std::string> NativeClient::get(const KeySpan &key) const {
+  return get(defaultColumnFamily(), key);
+}
+
+template <typename KeySpan>
+void NativeClient::del(const KeySpan &key) {
+  return del(defaultColumnFamily(), key);
+}
+
+inline WriteBatch NativeClient::getBatch() const { return WriteBatch{shared_from_this()}; }
+
+inline Iterator NativeClient::getIterator() const {
+  return std::unique_ptr<::rocksdb::Iterator>{client_->dbInstance_->NewIterator(::rocksdb::ReadOptions{})};
+}
+
+inline Iterator NativeClient::getIterator(const std::string &cFamily) const {
+  return std::unique_ptr<::rocksdb::Iterator>{
+      client_->dbInstance_->NewIterator(::rocksdb::ReadOptions{}, columnFamilyHandle(cFamily))};
+}
+
+inline std::vector<Iterator> NativeClient::getIterators(const std::vector<std::string> &cFamilies) const {
+  auto cfHandles = std::vector<::rocksdb::ColumnFamilyHandle *>{};
+  for (auto &cf : cFamilies) {
+    cfHandles.push_back(columnFamilyHandle(cf));
+  }
+  auto rawPtrIterators = std::vector<::rocksdb::Iterator *>{};
+  auto uniquePtrIterators = std::vector<std::unique_ptr<::rocksdb::Iterator>>{};
+  auto status = client_->dbInstance_->NewIterators(::rocksdb::ReadOptions{}, cfHandles, &rawPtrIterators);
+  // Are we supposed to delete the iterators if error is returned...
+  for (auto iter : rawPtrIterators) {
+    uniquePtrIterators.push_back(std::unique_ptr<::rocksdb::Iterator>{iter});
+  }
+  throwOnError("get multiple iterators failed"sv, std::move(status));
+  auto ret = std::vector<Iterator>{};
+  for (auto &iter : uniquePtrIterators) {
+    ret.push_back(std::move(iter));
```

Comment:
If you move this into the prior loop, you optimize for the positive case and only pay the penalty on the negative case. Otherwise we do a double loop on every successful call.

Reply:
You are right, thanks! Not sure why I used two loops, one loop is enough.